File: packages/flint-js/src/vegalite/templates/gantt.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { ChartTemplateDef } from '../../core/types';

/**
 * Gantt chart — one horizontal bar per task, spanning [start, end].
 *
 * Channels:
 *   - y      task / activity label (the banded category axis)
 *   - x      start of the interval (temporal or quantitative)
 *   - x2     end of the interval (shares x's scale)
 *   - color  optional grouping (phase, owner, resource)
 *
 * The interval lives on x/x2 so the bar is positioned by where it starts and
 * ends rather than measured from a zero baseline — hence markCognitiveChannel
 * 'position' and an explicit non-zero x scale. Tasks are sorted by their start
 * so the timeline reads top-to-bottom in chronological order.
 */
export const ganttChartDef: ChartTemplateDef = {
    chart: "Gantt Chart",
    template: {
        mark: { type: "bar", cornerRadius: 2, height: { band: 0.7 } },
        encoding: {},
    },
    channels: ["y", "x", "x2", "color", "detail", "column", "row"],
    markCognitiveChannel: 'position',
    declareLayoutMode: () => ({
        axisFlags: { y: { banded: true } },
    }),
    instantiate: (spec, ctx) => {
        const { x, x2, y, color, detail, column, row } = ctx.resolvedEncodings;

        if (!spec.encoding) spec.encoding = {};

        if (y) {
            spec.encoding.y = { ...y };
            spec.encoding.y.axis = { ...(spec.encoding.y.axis ?? {}), title: null };
            // Order tasks by when they start so the timeline reads chronologically.
            if (x?.field) {
                spec.encoding.y.sort = { field: x.field, op: 'min', order: 'ascending' };
            }
        }

        if (x) {
            spec.encoding.x = { ...x };
            spec.encoding.x.axis = { ...(spec.encoding.x.axis ?? {}), title: null };
            // A non-zero baseline only matters for a quantitative interval; on a
            // time scale Vega-Lite ignores (and warns about) scale.zero.
            if (x.type === 'quantitative') {
                spec.encoding.x.scale = { ...(spec.encoding.x.scale ?? {}), zero: false };
            }
        }
        // x2 shares x's scale; it only needs the field reference.
        if (x2) spec.encoding.x2 = { field: x2.field };

        if (color) spec.encoding.color = color;
        if (detail) spec.encoding.detail = detail;
        if (column) spec.encoding.column = column;
        if (row) spec.encoding.row = row;
    },
};

File: packages/flint-py/flint/vegalite/templates/gantt.py
"""Gantt Chart template — one horizontal bar per task, spanning [start, end].

Port of vegalite/templates/gantt.ts.
"""
from __future__ import annotations


def _gantt_declare_layout(cs, table, chart_properties):
    return {"axisFlags": {"y": {"banded": True}}}


def _gantt_instantiate(spec, ctx):
    encs = ctx["resolvedEncodings"]
    x = encs.get("x")
    x2 = encs.get("x2")
    y = encs.get("y")
    color = encs.get("color")
    detail = encs.get("detail")
    column = encs.get("column")
    row = encs.get("row")

    if "encoding" not in spec:
        spec["encoding"] = {}

    if y:
        spec["encoding"]["y"] = {**y}
        spec["encoding"]["y"]["axis"] = {**(spec["encoding"]["y"].get("axis") or {}), "title": None}
        # Order tasks by when they start so the timeline reads chronologically.
        if x and x.get("field"):
            spec["encoding"]["y"]["sort"] = {"field": x["field"], "op": "min", "order": "ascending"}

    if x:
        spec["encoding"]["x"] = {**x}
        spec["encoding"]["x"]["axis"] = {**(spec["encoding"]["x"].get("axis") or {}), "title": None}
        # A non-zero baseline only matters for a quantitative interval; on a time
        # scale Vega-Lite ignores (and warns about) scale.zero.
        if x.get("type") == "quantitative":
            spec["encoding"]["x"]["scale"] = {**(spec["encoding"]["x"].get("scale") or {}), "zero": False}

    # x2 shares x's scale; it only needs the field reference.
    if x2:
        spec["encoding"]["x2"] = {"field": x2["field"]} if x2.get("field") is not None else {}

    if color:
        spec["encoding"]["color"] = color
    if detail:
        spec["encoding"]["detail"] = detail
    if column:
        spec["encoding"]["column"] = column
    if row:
        spec["encoding"]["row"] = row


gantt_chart_def = {
    "chart": "Gantt Chart",
    "template": {
        "mark": {"type": "bar", "cornerRadius": 2, "height": {"band": 0.7}},
        "encoding": {},
    },
    "channels": ["y", "x", "x2", "color", "detail", "column", "row"],
    "markCognitiveChannel": "position",
    "declareLayoutMode": _gantt_declare_layout,
    "instantiate": _gantt_instantiate,
}
